libr/socket/run.c

```diff
@@ -1111,7 +1125,59 @@ R_API int r_run_start(RRunProfile *p) {
 		if (p->_pid) {
 			eprintf ("PID: Cannot determine pid with 'system' directive. Use 'program'.\n");
 		}
-		exit (r_sys_cmd (p->_system));
+		if (p->_daemon) {
+#if __WINDOWS__
+	//		eprintf ("PID: Cannot determine pid with 'system' directive. Use 'program'.\n");
+#else
+			pid_t child = r_sys_fork ();
+			if (child == -1) {
+				perror ("fork");
+				exit (1);
+			}
+			if (child) {
+				if (p->_pidfile) {
+					char pidstr[32];
+					snprintf (pidstr, sizeof (pidstr), "%d\n", child);
+					r_file_dump (p->_pidfile,
+							(const ut8*)pidstr,
+							strlen (pidstr), 0);
+				}
+				exit (0);
+			}
+			setsid ();
+			if (p->_timeout) {
+#if __UNIX__
+				int mypid = getpid ();
+				if (!r_sys_fork ()) {
+					int use_signal = p->_timeout_sig;
+					if (use_signal < 1) {
+						use_signal = SIGKILL;
+					}
+					sleep (p->_timeout);
+					if (!kill (mypid, 0)) {
+						// eprintf ("\nrarun2: Interrupted by timeout\n");
+					}
+					kill (mypid, use_signal);
+					exit (0);
+				}
+#else
+				eprintf ("timeout not supported for this platform\n");
+#endif
+			}
+#endif
+#if __UNIX__
+			close(0);
+			close(1);
+			exit (execl ("/bin/sh","/bin/sh", "-c", p->_system, NULL));
```

Comment:
Could you just call `r_sys_cmd` both here and in the `#else`? (and just move it outside of the `#if`)

Reply:
no. exec != system. one replaces the current process. the other creates a new. its needed to handle the pid properly